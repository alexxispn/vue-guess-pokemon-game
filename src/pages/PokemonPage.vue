<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else class="body">
    <h1>¿Quién es este Pokémon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions v-if="!showAnswer" :pokemons="pokemonArray" @selectedPokemon="checkAnswer"/>
    <template v-else>
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo Juego
      </button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOptions from "@/components/PokemonOptions";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  name: "PokemonPage",
  components: {PokemonOptions, PokemonPicture},
  data() {
    return {
      pokemonArray: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArray = await getPokemonOptions()
      const randomNumberFromOneToFour = () => Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArray[randomNumberFromOneToFour()]
    },
    checkAnswer(selectedId) {
      this.showPokemon = true
      this.showAnswer = true
      if (selectedId === this.pokemon.id) {
        this.message = `¡Correcto! Es ${this.pokemon.name}`
      } else {
        this.message = `¡Oops! Es ${this.pokemon.name}`
      }
    },
    newGame() {
      this.pokemonArray = []
      this.showPokemon = false
      this.showAnswer = false
      this.pokemon = null
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>

<style scoped>
h1 {
  font-family: 'Pokemon Solid', sans-serif;
  margin: 0;
  padding-top: 2rem;
}

.body {
  display: flex;
  flex-direction: column;
  align-items: center;
  height: 100vh;
  gap: 2rem;
}

</style>
