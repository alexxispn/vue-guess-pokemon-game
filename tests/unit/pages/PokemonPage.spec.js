import PokemonPage from "@/pages/PokemonPage";
import {shallowMount} from "@vue/test-utils";
import {PokemonsMock} from "../mocks/pokemons.mock";

describe("PokemonPage", () => {
    describe("before mount", () => {
        it("should call mixPokemonArray when mounted", () => {
            const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
            shallowMount(PokemonPage)
            expect(mixPokemonArraySpy).toHaveBeenCalled()
        })
    })
    describe(" after mount", () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArray: PokemonsMock,
                    pokemon: PokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        const PokemonPicture = wrapper.find('pokemon-picture-stub')
        const PokemonOptions = wrapper.find("pokemon-options-stub")
        it('should match with snapshot after pokemons are loaded', () => {
            expect(wrapper.html()).toMatchSnapshot()
        })
        it("should render PokemonPicture & PokemonOptions components", () => {
            expect(PokemonPicture.exists).toBeTruthy()
            expect(PokemonOptions.exists).toBeTruthy()
        })
        it('should be pokemonId 1 in PokemonPicture', () => {
            expect(PokemonPicture.attributes('pokemonid')).toBe('1')
        })
        describe('checkAnswer', () => {
            it('should return correct message if the answer is correct', async () => {
                await wrapper.vm.checkAnswer(1)
                const h2 = wrapper.find('h2')
                expect(h2.exists).toBeTruthy()
                expect(wrapper.vm.showPokemon).toBeTruthy()
                expect(h2.text()).toBe(`¡Correcto! Es ${PokemonsMock[0].name}`)
            })
            it('should return error message if the answer is not correct', async () => {
                await wrapper.vm.checkAnswer(2)
                const h2 = wrapper.find('h2')
                expect(h2.exists).toBeTruthy()
                expect(h2.text()).toBe(`¡Oops! Es ${PokemonsMock[0].name}`)
            })
        })
    })
})
