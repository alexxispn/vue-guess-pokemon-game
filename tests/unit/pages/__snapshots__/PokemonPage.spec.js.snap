// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage  after mount should match with snapshot after pokemons are loaded 1`] = `
<div class="container">
  <h1 class="title">¿Cuál es ese Pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
</div>
`;
